### PR TITLE
Fix for issue #18

### PR DESCRIPTION
**Pull request by catholicon**

### CODE REVIEW

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java [hunk 1]
```diff
@@ -28,10 +30,20 @@
 public class ClassRuleUtils {
     public static final Logger LOG = LoggerFactory.getLogger(ClassRuleUtils.class);
 
+    /**
+     * If index lanes should be dynamically detected
+     */
+    protected static final boolean DETECT_INDEXING_LANES = Boolean.getBoolean("cq.testing.detectIndexingLanes");
```
Comment (volteanu) [inline src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: Let's align this with the other and call it `it.detectIndexingLanes`. Also, I suggest moving the reading of the sys prop to a function similar to `loginTokenAuth()` for testability.

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java [hunk 2]
```diff
@@ -28,10 +30,20 @@
...
+    /**
+     * Indexing lanes to use if not detected dynamically
+     */
+    protected static final String[] DEFAULT_INDEXING_LANES = new String[]{"async", "fulltext-async"};
+
     // TODO way to make these props homogenous?
     public static final String LOGIN_TOKEN_AUTH = "it.logintokenauth";
 
-    public static class LoginTokenInstance extends ExistingInstance {
+    public static class LoginTokenInstance extends ConfigureIndexingLaneInstance {
```
Comment (volteanu) [inline src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: The chained inheritance doesn't seem like a good pattern here. What about we make a `ConfigurableInstance` class (in its own file 😄) that handles all the system props? Having the logic grouped in one place will help to handle all the combinations of flags.

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java [hunk 3]
```diff
@@ -41,6 +53,18 @@
         }
     }
 
+    public static class ConfigureIndexingLaneInstance extends ExistingInstance {
+        @Override
+        public <T extends SlingClient> T getClient(Class<T> clientClass, String user, String pass) {
```
Comment (volteanu) [inline src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: You should actually override `newClient` and `newAdminClient` which are used by `get[Admin]Client`.

**Review by volteanu — CHANGES_REQUESTED** [review]

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java (hunk 2 above)
Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: .... well, I can make `ConfigurableInstance` as a separate but afaict, `ClassRuleUtils#LoginTokenInstance` would still need to inherit from that. Similarly I guess `ClassRuleUtils#newInstanceRule` would just switch between `LoginTokenInstance` or `ConfigurableInstance`.

Am I missing something?

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java (hunk 3 above)
Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: done in catholicon/aem-testing-clients@ebad5e5

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java (hunk 1 above)
Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: done in catholicon/aem-testing-clients@ebad5e5

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java (hunk 2 above)
Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: I've moved `ConfiguredIndexLanesInstance` in a separate class at catholicon/aem-testing-clients@ebad5e5. We can, of course, discuss the naming (I don't want to drop `IndexLanes` from it - configured or configurable is probably up for discussion... since the lanes are hard-coded I would prefer "configured".

Comment (volteanu) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: I was thinking we can drop `LoginTokenInstance` and use only `ConfigurableInstance` as a replacement, which would handle both login and index lanes configs.

Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: Hmm... Yes, that maked sense. But, I think, that should be a separate issue. I can open one if you want.

I'm guessing then that the current flag for basic auth would be overloaded as use both basic auth and injected lanes. Or, maybe, you meant that we should have two different booleans and hence probably another `newInstanceRule`?

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java [hunk 4]
```diff
@@ -31,7 +30,7 @@
     // TODO way to make these props homogenous?
     public static final String LOGIN_TOKEN_AUTH = "it.logintokenauth";
 
-    public static class LoginTokenInstance extends ExistingInstance {
+    public static class LoginTokenInstance extends ConfiguredIndexingLaneInstance {
```
Comment (dulvac) [inline src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: I don't think I like this. Can't we just have a separate `Rule` with a separate statement? `LoginTokenInstance` does not logically depend on indexing. We can link them together in the `CQAuthorRule`s top-level convenience rules.

**Review by dulvac — CHANGES_REQUESTED** [review]

File: src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java (hunk 4 above)
Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: @dulvac, I'm sorry I'm not quite sure I understand your suggestion. Please consider me a complete novice here. Could you please guide me towards how exactly do you want this change to be implemented? (btw, as you might notice in earlier conversation - I don't want to invest time in refactoring other pieces of code for this change - that, imvho, should be taken up as a separate issue and done by someone well versed)

Comment (dulvac) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: @catholicon , no worries. What I mean by this is I think we have two options:

A: If it's a corner case and only affect certain tests
* `LoginTokenInstance` should NOT extend `ConfiguredIndexingLaneInstance` 
* `ConfiguredIndexingLaneInstance` should NOT extend `ExistingInstance`.
* Maybe we have a separate `ClassRule` which takes an `Instance` in the constructor, that you need to explicitly add to your test and call `setLaneNames` on the default client

B: if we want to enable it for _every_ test by default
* What @volteanu said, and I support it:
* Rename `LoginTokenInstance` to `ConfigurableInstance` and add your logic there.

Comment (catholicon) [reply src/main/java/com/adobe/cq/testing/junit/rules/ClassRuleUtils.java]: So, something like catholicon/aem-testing-clients@3bb4d16 maybe?

**Review by volteanu — APPROVED** [review]
That looks perfect!
